doc/formatter.md

```diff
@@ -67,6 +67,79 @@ subproblem: _Given a box width constraint, what are the best ways to format a
+![Inheritance diagram](./formatter_simplified_class_diagram.png)
+
+[TreeUnwrapper] class exploits
+the [Visitor](https://github.com/google/verible/blob/ff1c692695ca723275f24397572cb1d1eedb6779/common/text/visitors.h#L44)
+pattern to recursively process the input CST tree:
+```cpp
+// SymbolVisitor is an abstract visitor class from which visitors can be derived
+// SymbolVisitor only visits a single Symbol, and does not recurse on nodes
+//
+// Usage:
+//  SymbolPtr tree = ...
+//  SymbolVisitorSubclass visitor;
+//  tree->Accept(visitor);
+//
+class SymbolVisitor {
+ public:
+  virtual ~SymbolVisitor() {}
+  virtual void Visit(const SyntaxTreeLeaf& leaf) = 0;
+  virtual void Visit(const SyntaxTreeNode& node) = 0;
+};
+```
+
+It builds the [TokenPartitionTree] (i.e. `VectorTree<UnwrappedLine>`) with a set of helper methods to simplify the process:
+* [`TraverseChildren()`](https://github.com/google/verible/blob/ff1c692695ca723275f24397572cb1d1eedb6779/common/formatting/tree_unwrapper.cc#L278) - only traverses the CST
+* [`VisitIndentedSection()`](https://github.com/google/verible/blob/ff1c692695ca723275f24397572cb1d1eedb6779/common/formatting/tree_unwrapper.cc#L324) - creates indented section and traverses the CST
+* `VisitNewUnwrappedLine()` - starts new line (in current indented section) and traverses the CST
+
+The following two code blocks show examples of such:
+```cpp
+void TreeUnwrapper::TraverseChildren(const verible::SyntaxTreeNode& node) {
+  // Can't just use TreeContextVisitor::Visit(node) because we need to
+  // call a visit hook between children.
+  const verible::SyntaxTreeContext::AutoPop p(&current_context_, &node);
+  InterChildNodeHook(node);
+  for (const auto& child : node.children()) {
+    if (child) {
+      child->Accept(this);
+      InterChildNodeHook(node);
+    }
+  }
+}
+```
+
+```cpp
+void TreeUnwrapper::VisitIndentedSection(const SyntaxTreeNode& node,
+                                         int indentation_delta,
+                                         PartitionPolicyEnum partitioning) {
+  const auto last_ftoken_iter =
+      VisitIndentedChildren(node, indentation_delta, partitioning);
+  // TODO(fangism): do we ever need to remove trailing empty partitions here?
+  // Update parent's end() format token iterator to match that of
+  // its last child.  It can still be advanced later.
+  active_unwrapped_lines_->Value().SpanUpToToken(last_ftoken_iter);
+  // Start new empty UnwrappedLine at the previous indentation level.
+  StartNewUnwrappedLine(PartitionPolicyEnum::kUninitialized, nullptr);
+}
+```
+
+### Custom policies application
+
+The next stage of code formatting is to apply custom functions that would
+reshape the partition tree according to the selected policy.
+An example of that are the [kTabularAlignment](#alignment) and the
+[kAppendFittingSubPartitions](https://cs.opensource.google/verible/verible/+/master:common/formatting/unwrapped_line.h;bpv=1;bpt=1;l=65)
```

Comment:
Try the `symbol:` search atom.

Reply:
I'm keeping it this way as search engine can't find such symbol (which is weird because it does find enum constants in abseil code).